### PR TITLE
New C++ API: Add checks for Sorts.

### PR DESCRIPTION
**Pull request by aniemetz**

### CODE REVIEW

File: src/expr/type.h [hunk 1]
```diff
@@ -144,6 +144,32 @@ class CVC4_PUBLIC Type {
...
+  /**
+   * Is this a function-LIKE type?
+   *
+   * Function-like things (e.g. datatype selectors) that aren't actually
+   * functions ARE considered functions, here. The main point is that this is
+   * used to avoid anything higher-order: anything function-like can not be
+   * the argument or return value for anything else function-like.
+   *
+   * Arrays are explicitly NOT function-like for the purposes of this test.
+   * However, functions still can not contain anything function-like.
```
Comment (4tXJ7f) [inline src/expr/type.h]: I am not sure that the last sentence is needed (it is also not in the API header).

File: src/api/cvc4cpp.h [hunk 2]
```diff
@@ -348,7 +382,118 @@ class CVC4_PUBLIC Sort
...
+   * @return the bit-width of the bit-vector sort
+   */
+  unsigned getBVSize() const;
+
+  /* Floating-point sort ------------------------------------------------- */
+
+  /**
+   * @return the bit-width of the floating-point exponent sort
+   */
+  unsigned getFPExponentSize() const;
```
Comment (4tXJ7f) [inline src/api/cvc4cpp.h]: Given that the size of a bit-vector is determined by a `uint32_t`, this should return a `uint32_t` as well (IMHO).

File: src/api/cvc4cpp.h [hunk 3]
```diff
@@ -348,7 +382,118 @@ class CVC4_PUBLIC Sort
...
+
+  /**
+   * @return the width of the floating-point significand sort
+   */
+  unsigned getFPSignificandSize() const;
```
Comment (4tXJ7f) [inline src/api/cvc4cpp.h]: Given that the size of a bit-vector is determined by a `uint32_t`, this should return a `uint32_t` as well (IMHO).

File: src/api/cvc4cpp.cpp [hunk 4]
```diff
@@ -2423,9 +2613,25 @@ Term Solver::defineFun(Term fun,
...
+  CVC4_API_ARG_SIZE_CHECK_EXPECTED(
+      size == domain_sorts.size(), bound_vars, domain_sorts.size());
+  for (size_t i = 0; i < size; ++i)
+  {
+    CVC4_API_ARG_AT_INDEX_CHECK_EXPECTED(
+        domain_sorts[i] == bound_vars[i].getSort(),
+        "sort of parameter",
+        bound_vars[i],
+        i,
+        "'" << domain_sorts[i] << "'");
```
[the remark this answers is not in the document]
Comment (aniemetz) [reply src/api/cvc4cpp.cpp]: That's overkill IMO. All Macros that wrap CVC4_API_CHECK do not support this - would you then want all of them to support this?

File: src/api/cvc4cpp.h [hunk 5]
```diff
@@ -348,7 +382,118 @@ class CVC4_PUBLIC Sort
...
+
+  /**
+   * @return the bit-width of the bit-vector sort
+   */
+  unsigned getBVSize() const;
+
+  /* Floating-point sort ------------------------------------------------- */
+
+  /**
+   * @return the bit-width of the floating-point exponent sort
```
[the remark this answers is not in the document]
Comment (aniemetz) [reply src/api/cvc4cpp.h]: That's what it says: "exponent sort". But I can reword nonetheless.

File: src/api/cvc4cpp.h [hunk 6]
```diff
@@ -348,7 +382,118 @@ class CVC4_PUBLIC Sort
...
+   */
+  unsigned getFPExponentSize() const;
+
+  /**
+   * @return the width of the floating-point significand sort
```
[the remark this answers is not in the document]
Comment (aniemetz) [reply src/api/cvc4cpp.h]: Ditto

**Pull request merged**